analytics/transform_test.go

```diff
@@ -222,17 +220,30 @@ func BenchmarkToJson(b *testing.B) {
 	}
 }
 
-func TestToJsonWithGeo(t *testing.T) {
-	assert := assert.New(t)
-
-	// correct value
-	jsonEvent, err := stdJson.Marshal(eventMapWithGeo)
+// orderJson uses the standard json library to remarshal a []byte, resulting in a sorted JSON
+func orderJson(bytes []byte) ([]byte, error) {
```

Comment:
Continuing the conversation from here: https://github.com/snowplow/snowplow-golang-analytics-sdk/pull/21#discussion_r869051900

> Yes, this is only for the purpose of testing. It is not used to make it more human-readable, it is used instead to be able to assert the equality of the entire Jsons, as they are not ordered anymore now that we use jsoniter without key sorting and having both sorted gets us identical objects.

So in light of this reasoning, I don't see why we would unmarshal and then marshal the test value for this purpose.

We have an expected value which we hardcode, and a test value which should always have the keys in the same order. Instead of doing all this work to reorder the keys, and introducing extra scope for error, why don't we just set the expected value's keys in the order we expect them?

Reply:
The issue here lies in the fact that the json produced by ToJsonWithGet (or any of the other JSON-producing functions) is not only unsorted, it is also random. Here is the output from two different runs of ToJsonWithGeo:

```
{"etl_tstamp":"2013-11-26T00:03:57.885Z","name_tracker":"cloudfront-1","user_ipaddress":"92.231.54.234","page_url":"http://www.snowplowanalytics.com","page_title":"On Analytics","contexts_com_snowplowanalytics_snowplow_ua_parser_context_1":[{"osFamily":"Windows XP","osVersion":"Windows XP","osPatch":null,"useragentFamily":"IE","useragentMajor":"7","useragentMinor":"0","useragentPatch":null,"useragentVersion":"IE 7.0","osMajor":null,"osMinor":null,"osPatchMinor":null,"deviceFamily":"Other"}],"collector_tstamp":"2013-11-26T00:03:57.885Z","v_collector":"clj-tomcat-0.1.0","geo_zipcode":"94109","page_urlport":80,"page_urlpath":"/product/index.html","page_urlfragment":"4-conclusion","unstruct_event_com_snowplowanalytics_snowplow_link_click_1":{"targetUrl":"http://www.example.com","elementClasses":["foreground"],"elementId":"exampleLink"},"event_id":"c6ef3124-b53a-4b13-a233-0088f79dcbcb","txn_id":41828,"network_userid":"ecdff4d0-9175-40ac-a8bb-325c49733607","geo_city":"New York","page_urlscheme":"http","contexts_org_schema_web_page_1":[{"genre":"blog","inLanguage":"en-US","datePublished":"2014-11-06T00:00:00Z","author":"Fred Blundun","breadcrumb":["blog","releases"],"keywords":["snowplow","javascript","tracker","event"]}],"br_features_pdf":true,"domain_sessionid":"2b15e5c8-d3b1-11e4-b9d6-1681e6b88ec1","v_etl":"serde-0.5.2","geo_region_name":"Florida","page_urlhost":"www.snowplowanalytics.com","contexts_org_w3_performance_timing_1":[{"domContentLoadedEventEnd":1415358091309,"navigationStart":1415358089861,"unloadEventStart":1415358090270,"redirectStart":0,"redirectEnd":0,"fetchStart":1415358089870,"connectEnd":1415358090183,"domLoading":1415358090270,"domainLookupStart":1415358090102,"domainLookupEnd":1415358090102,"connectStart":1415358090103,"responseEnd":1415358090265,"domContentLoadedEventStart":1415358090968,"domComplete":0,"loadEventStart":0,"unloadEventEnd":1415358090287,"responseStart":1415358090265,"domInteractive":1415358090886,"requestStart":1415358090183,"loadEventEnd":0}],"derived_tstamp":"2013-11-26T00:03:57.885Z","event_version":"1-0-0","event_fingerprint":"e3dbfa9cca0412c3d4052863cefb547f","v_tracker":"js-2.1.0","user_id":"jon.doe@email.com","geo_region":"TX","geo_longitude":-122.4124,"event_vendor":"com.snowplowanalytics.snowplow","event_name":"link_click","true_tstamp":"2013-11-26T00:03:57.885Z","geo_location":"37.443604,-122.4124","app_id":"\u003c\u003eangry-birds","platform":"web","event":"page_view","domain_sessionidx":3,"geo_latitude":37.443604,"ip_isp":"FDN Communications","ip_organization":"Bouygues Telecom","ip_domain":"nuvox.net","br_features_flash":false,"dvce_created_tstamp":"2013-11-26T00:03:57.885Z","domain_userid":"bc2e92ec6c204a14","geo_country":"US","ip_netspeed":"Cable/DSL","event_format":"jsonschema","user_fingerprint":"2161814971","page_urlquery":"id=GTM-DLRG"}
```
```
{"geo_region_name":"Florida","ip_netspeed":"Cable/DSL","page_urlquery":"id=GTM-DLRG","page_urlfragment":"4-conclusion","dvce_created_tstamp":"2013-11-26T00:03:57.885Z","domain_userid":"bc2e92ec6c204a14","geo_country":"US","geo_latitude":37.443604,"event_name":"link_click","event_fingerprint":"e3dbfa9cca0412c3d4052863cefb547f","true_tstamp":"2013-11-26T00:03:57.885Z","geo_city":"New York","ip_organization":"Bouygues Telecom","page_urlpath":"/product/index.html","domain_sessionid":"2b15e5c8-d3b1-11e4-b9d6-1681e6b88ec1","unstruct_event_com_snowplowanalytics_snowplow_link_click_1":{"elementClasses":["foreground"],"elementId":"exampleLink","targetUrl":"http://www.example.com"},"br_features_pdf":true,"br_features_flash":false,"event_version":"1-0-0","app_id":"\u003c\u003eangry-birds","user_id":"jon.doe@email.com","user_fingerprint":"2161814971","domain_sessionidx":3,"derived_tstamp":"2013-11-26T00:03:57.885Z","txn_id":41828,"geo_region":"TX","geo_zipcode":"94109","contexts_org_w3_performance_timing_1":[{"responseStart":1415358090265,"domInteractive":1415358090886,"domainLookupEnd":1415358090102,"requestStart":1415358090183,"responseEnd":1415358090265,"domComplete":0,"navigationStart":1415358089861,"unloadEventStart":1415358090270,"redirectStart":0,"domainLookupStart":1415358090102,"loadEventEnd":0,"domLoading":1415358090270,"domContentLoadedEventEnd":1415358091309,"redirectEnd":0,"fetchStart":1415358089870,"connectStart":1415358090103,"connectEnd":1415358090183,"unloadEventEnd":1415358090287,"domContentLoadedEventStart":1415358090968,"loadEventStart":0}],"page_urlhost":"www.snowplowanalytics.com","contexts_org_schema_web_page_1":[{"breadcrumb":["blog","releases"],"keywords":["snowplow","javascript","tracker","event"],"genre":"blog","inLanguage":"en-US","datePublished":"2014-11-06T00:00:00Z","author":"Fred Blundun"}],"event_vendor":"com.snowplowanalytics.snowplow","geo_location":"37.443604,-122.4124","etl_tstamp":"2013-11-26T00:03:57.885Z","collector_tstamp":"2013-11-26T00:03:57.885Z","v_tracker":"js-2.1.0","geo_longitude":-122.4124,"page_url":"http://www.snowplowanalytics.com","page_title":"On Analytics","contexts_com_snowplowanalytics_snowplow_ua_parser_context_1":[{"useragentFamily":"IE","useragentMajor":"7","useragentMinor":"0","useragentPatch":null,"useragentVersion":"IE 7.0","osFamily":"Windows XP","osMajor":null,"osMinor":null,"osPatchMinor":null,"osVersion":"Windows XP","deviceFamily":"Other","osPatch":null}],"event":"page_view","event_id":"c6ef3124-b53a-4b13-a233-0088f79dcbcb","v_etl":"serde-0.5.2","user_ipaddress":"92.231.54.234","ip_domain":"nuvox.net","page_urlscheme":"http","event_format":"jsonschema","name_tracker":"cloudfront-1","v_collector":"clj-tomcat-0.1.0","network_userid":"ecdff4d0-9175-40ac-a8bb-325c49733607","ip_isp":"FDN Communications","platform":"web","page_urlport":80}
```

Reply:
OK, so I see what happened here - it was sorting it before, but when we removed the config it no longer sorts. 

OK, cool, we can leave the sorting in the test so.